packages/lib/src/vis/hooks.ts

```diff
@@ -54,23 +54,20 @@ export function useDomains(
 }
 
 export function useCameraState<T>(
-  factory: (camera: Camera, context: VisCanvasContextValue) => T,
-  deps: unknown[]
+  factory: (camera: Camera) => T,
+  equalityFn?: (prev: T, next: T) => boolean
 ): T {
   const camera = useThree((state) => state.camera);
-  const context = useVisCanvasContext();
 
-  const [state, setState] = useState(() => factory(camera, context));
+  const factoryRef = useSyncedRef(factory);
+  const [state, setState] = useState(() => factoryRef.current(camera));
 
-  function updateState() {
-    setState(() => factory(camera, context)); // functional update to skip re-renders if state hasn't changed
-  }
-
-  // Update state when context value or dependencies change
-  useEffect(updateState, [camera, context, ...deps]); // eslint-disable-line react-hooks/exhaustive-deps
-
-  // Update state when canvas is redrawn (i.e. on zoom/pan)
-  useFrame(updateState);
+  useFrame(() => {
+    const next = factoryRef.current(camera);
+    if (!equalityFn || !equalityFn(state, next)) {
+      setState(next);
```

Comment:
In case this isn't clear in the docs, when `equalityFn` is not set, I rely on React to perform a strict equality check. The comment I had  before was wrong: `setState` always performs an equality check regardless of how you call it.